src/Workspaces/Remote/Core/Services/AssetService.cs

```diff
@@ -172,6 +239,88 @@ public Entry(object @object)
+                var solutionChecksumObject = await _assetService.GetAssetAsync<SolutionChecksumObject>(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                AppendIfNeeded(solutionChecksums, solutionChecksumObject.Info);
+                AppendIfNeeded(solutionChecksums, solutionChecksumObject.Projects);
+
+                await _assetService.SynchronizeAssetsAsync(solutionChecksums, cancellationToken).ConfigureAwait(false);
+
+                // get children of project checksum objects at once
+                var projectChecksums = new HashSet<Checksum>();
+                foreach (var projectChecksum in solutionChecksumObject.Projects)
+                {
+                    var projectChecksumObject = await _assetService.GetAssetAsync<ProjectChecksumObject>(projectChecksum, cancellationToken).ConfigureAwait(false);
+
+                    AppendIfNeeded(projectChecksums, projectChecksumObject.Info);
+                    AppendIfNeeded(projectChecksums, projectChecksumObject.CompilationOptions);
+                    AppendIfNeeded(projectChecksums, projectChecksumObject.ParseOptions);
+
+                    AppendIfNeeded(projectChecksums, projectChecksumObject.ProjectReferences);
+                    AppendIfNeeded(projectChecksums, projectChecksumObject.MetadataReferences);
+                    AppendIfNeeded(projectChecksums, projectChecksumObject.AnalyzerReferences);
+                    AppendIfNeeded(projectChecksums, projectChecksumObject.Documents);
+                    AppendIfNeeded(projectChecksums, projectChecksumObject.AdditionalDocuments);
+                }
+
+                await _assetService.SynchronizeAssetsAsync(projectChecksums, cancellationToken).ConfigureAwait(false);
+
+                // get children of document checksum objects at once
+                var documentChecksums = new HashSet<Checksum>();
+                foreach (var projectChecksum in solutionChecksumObject.Projects)
+                {
+                    var projectChecksumObject = await _assetService.GetAssetAsync<ProjectChecksumObject>(projectChecksum, cancellationToken).ConfigureAwait(false);
+
+                    foreach (var checksum in projectChecksumObject.Documents)
+                    {
+                        var documentChecksumObject = await _assetService.GetAssetAsync<DocumentChecksumObject>(checksum, cancellationToken).ConfigureAwait(false);
+
+                        AppendIfNeeded(documentChecksums, documentChecksumObject.Info);
+                        AppendIfNeeded(documentChecksums, documentChecksumObject.Text);
+                    }
+
+                    foreach (var checksum in projectChecksumObject.AdditionalDocuments)
+                    {
+                        var documentChecksumObject = await _assetService.GetAssetAsync<DocumentChecksumObject>(checksum, cancellationToken).ConfigureAwait(false);
+
+                        AppendIfNeeded(documentChecksums, documentChecksumObject.Info);
```

Comment:
i worrk about how much knowledge has to be everywhere about which parts of an object there are that need to sync.  It would be nice if there was just some interface for tehse objects that exposes that, so you could write this sort of code all over the place in a generic fashion.  This is especially important if someone ends up adding new data to a Workspace entity in the future.  Knowing where/what/how to update all this code is not clear at all.